File: leetcode/00200-00299/00278_first-bad-version/Solution.py
#!/usr/bin/env python3
# -*- coding:UTF-8 -*-

def isBadVersion(n):
    if n>= 4:
        return True
    else :
        return False
        
class Solution:

    # Solution 1
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        start = 1
        end = n
        while start <= end:
            middle = (start + end)>>1
            if isBadVersion(middle):
                end = middle - 1
            else:
                start = middle + 1
        return start
        
if __name__ == '__main__':
    solution = Solution()
    print(solution.firstBadVersion(500))

File: leetcode/00200-00299/00278_first-bad-version/Solution.java

class Solution extends VersionControl {
    
    
    public int firstBadVersion(int n) {
        int start = 1;
        int end = n;
        int pos = 0;
        while ( start <= end ) {
            int middle = start + ((end - start)>>1);
            if ( isBadVersion(middle) ) {
                pos = middle;
                end = middle - 1;
            } else {
                start = middle + 1;
            }
        }
        return pos;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int result = solution.firstBadVersion(500);
        System.out.println(result);
    }
}